Feature/modbus data type

hutcheb opened the pull request:
Add support for BOOL, INT, UINT and REAL data types for the Modbus protocol.
address string can now include any of these types such as:- holding-register:10:REAL[10], extended-register:10:UINT[10].
Note the previous format now defaults to BOOL or INT depending on the memory area.

Add support to parse values from Java Strings to any of these types. This fixes a bug in the hello-world example when writing.

Add support for BOOL, INT, UINT and REAL PlcValue types. This have been added as we are able to include the max/min value checks and casting of the input type within the constructors cleaning up the ModbusFieldHandler logic. The method getBytes has also been added to clean up the ModbusProtocolLogic class.

I'll add Documentation once the other Modbus Documentation PR has been finalized.

hutcheb committed:
Cast to PLCValue within data-io-template

Add a cast to the PLCValue parameter so that we are able to use non 
standard methods to decode request

hutcheb committed:
Modbus.spec Sample Change

Sample Modbus.mspec change to be able to specify the datatype when 
parsing the PLCValue.

hutcheb committed:
Add Datatype to ModbusField classes

hutcheb committed:
Add sample PLCValue classes based on IEC 61131 datatypes.

This allows IEC61131 datatypes INT (r/w) and UINT (Write only) to be 
used to convert Modbus registers.

hutcheb committed:
Tidied up the IEC type classes INT and UINT.

hutcheb committed:
Fixed support for PLCLists for INT and UINT datatypes

hutcheb committed:
Cleaned up, it now supports BOOL, INT, UINT and REAL values.

hutcheb committed:
Test cases for some Encoding and Modbus Field methods.

niclash commented:
I have also seen 32-bit integers. Maybe it should be available too.

hutcheb committed:
Add support for DINT data types

hutcheb commented:
Hi Niclas,

The plan is to add support for as many IEC 61131 data types as possible. I've added DINT as an example for what it would take to add additional data types but will wait for more feedback before adding more.

Ben

hutcheb committed:
Included cast in data-io-template for value in if statement.
This is to avoid needlessly casting it.

hutcheb committed:
Added test cases for negative REAL numbers.

chrisdutz commented on plc4j/api/src/main/java/org/apache/plc4x/java/api/value/PlcBOOL.java:
```diff
@@ -0,0 +1,214 @@
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, property = "className")
+public class PlcBOOL extends PlcIECValue<Boolean> {
+
+    BigInteger minValue = BigInteger.valueOf(0);
+    BigInteger maxValue = BigInteger.valueOf(1);
```
Could we perhaps make these static ... A BigInteger is a pretty big type and creating these for every boolean PlcValue seems wasting memory and CPU time for cleaning up.

chrisdutz commented on plc4j/api/src/main/java/org/apache/plc4x/java/api/value/PlcDINT.java:
```diff
@@ -0,0 +1,103 @@
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, property = "className")
+public class PlcDINT extends PlcIECValue<Integer> {
+
+    BigInteger minValue = BigInteger.valueOf((long) Integer.MIN_VALUE);
+    BigInteger maxValue = BigInteger.valueOf((long) Integer.MAX_VALUE);
```
Same as with PlcBOOL (And probably all others)

chrisdutz reviewed (approved):
I really like this PR ... I guess it would simplify driver implementation quite a bit ... the thing with the BigIntegers as instance variables we can always change.

chrisdutz commented:
Do you need any help from my side in order to continue? If yes, just reach out to me.

hutcheb commented:
I have created all of the numeric datatypes and char and wchar types as well as updated the ModbusFieldHandler file.

I'm just sorting out some issues this morning with how BigIntegers are converted to byte[] within the PlcULINT and PlcLWORD classes.

By the end of my session this morning I should be in a position to start writing test cases in the Modbus section.

I haven't touched the other drivers.

I'll commit what I have when I've finished this morning next 15mins or so, feel free to carry on as I won't be touching it until tomorrow.

hutcheb committed:
Added most of the other datatypes and support in Modbus

hutcheb committed:
Modbus now supports all numeric and string and char datatypes

Outstanding items:-
There is a readBigInteger method which isn't implemented yet. Exception raised when reading ULINT.
Documentation
Discuss if this should be included in other protocols at this point.

hutcheb committed:
Feature/Modbus Datatype - IEC 61131

Allow IEC 61131 values to be use with the Modbus protocol

hutcheb commented:
Hi Chris,

This should be in a position for you to review.

Ben

hutcheb committed:
Change the data-io template to use DataType Instead of List

This was done so that a PLCList can be created from data from the PLC of 
the IEC types. Otherwise Java PlcValue would be used.

hutcheb committed:
Tidying up.

chrisdutz commented:
Hi Ben,

will do so asap …

Chris

Von: hutcheb <notifications@github.com>
Antworten an: apache/plc4x <reply@reply.github.com>
Datum: Samstag, 19. September 2020 um 13:30
An: apache/plc4x <plc4x@noreply.github.com>
Cc: Christofer Dutz <christofer.dutz@c-ware.de>, Comment <comment@noreply.github.com>
Betreff: Re: [apache/plc4x] Feature/modbus data type (#186)


Hi Chris,

This should be in a position for you to review.

Ben

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/apache/plc4x/pull/186#issuecomment-695201928>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAE66YL4X6FWUSX3B2I2RYLSGSI5RANCNFSM4REYHRRQ>.

hutcheb commented:
Hi,

I was looking at starting to use the staticSerialize method for Modbus as it would reduce the fromPLCValue method. However there are a few methods in the WriteBuffer class that aren't implemented. writeDouble and the writeBigInteger functions. The writeDouble function in particular would cause issues with writing floats of any kind for all protocols that use it.

Does writing floats with the S7 protocol work or am I missing something?

hutcheb committed:
Merge branch 'develop' of https://github.com/apache/plc4x into Feature/ModbusDataType

chrisdutz commented:
Hi Ben ... sorry for the late response ... I'm currently drowning in getting some stuff finished as well as preparing for ApacheCon next week ... I'll try to squeeze it in, but can't promise ... haven't forgotten you and your hard work ... just want to invest the time for reviewing, that your work deserves.

hutcheb commented:
Hi Chris, not a problem take your time. :)

chrisdutz commented:
The PR #190 now contains all changes of this PR.
